backend/internal/core/models/voucher.go

```diff
@@ -6,10 +6,13 @@ import (
 )
 
 var ErrVoucherNotFound = fmt.Errorf("voucher is not found")
+var ErrVoucherRedeemed = fmt.Errorf("unredeemed voucher with this code is not found")
 
 // Voucher struct holds all data for vouchers, voucher used only by one user.db.
 type Voucher struct {
 	ID        int       `json:"id" gorm:"primaryKey;autoIncrement"`
+	UserID    int       `json:"user_id"`
+	Username  string    `json:"username"`
```

Comment:
I think the denormalization here is fine, while the username be wrong in the future if the user decide to change their username, it doesn't matter